[WIP] Integrate cuSpatial with gpuCI

dillon-cullinan opened the pull request:
Depends on: https://github.com/rapidsai/cuspatial/pull/39

- [x] GPU build.sh script
- [x] CPU build.sh scripts
- [x] Changelog check script
- [x] Style check script
- [x] Conda meta.yaml
- [x] Conda build.sh script
- [x] gpuci-scripts job updates https://github.com/rapidsai/gpuci-scripts/pull/60
- [x] Internal non-GPU testing

dillon-cullinan commented:
At this point I have a strong outline of the basis for all the scripts. I would like @harrism to review some of the conda stuff. `cudf` has many header files that it tests in its `meta.yaml`, should I be including all of cuspatials header files in here similarly?

The dependencies are all the same as cudf right now since you currently require a cudf source build. Once you are able to build with a cudf conda package we can remove a lot of this. Reliance on the cudf source build also will heavily increase CI build times.

These changes are all untested for now. Will be focused on testing CPU builds tomorrow.

raydouglass commented on ci/cpu/prebuild.sh:
```diff
@@ -0,0 +1,15 @@
+
+#Upload cuspatial once per CUDA
+if [[ "$CUDA" == "9.2" ]]; then
+    export UPLOAD_CUDF=1
```
Rename these to `UPLOAD_CUSPATIAL`/`UPLOAD_LIBCUSPATIAL`

raydouglass commented on ci/gpu/build.sh:
```diff
@@ -0,0 +1,99 @@
+
+logger "Activate conda env..."
+source activate gdf
+conda install "rmm=$MINOR_VERSION.*" "cudatoolkit=$CUDA_REL" \
```
I'm assuming the conda & pip installs are for cudf, not cuspatial? Can you document that so when a from-source build of cudf isn't necessary, these installs can be removed?

raydouglass commented on ci/gpu/build.sh:
```diff
@@ -0,0 +1,99 @@
+################################################################################
+
+logger "Clone cudf"
+git clone git@github.com:rapidsai/cudf.git ${CUDF_HOME}
```
You probably want specify which branch to clone: `git clone git@github.com:rapidsai/cudf.git -b branch-$MINOR_VERSION ${CUDF_HOME}`

raydouglass commented on conda/recipes/libcuspatial/build.sh:
```diff
@@ -0,0 +1,14 @@
+git clean -xdf
+# build cudf with verbose output
+cd $CUDF_HOME
+./build.sh -v libnvstrings nvstrings libcudf cudf
```
Does `libcuspatial` need `nvstrings` and `cudf` and not just the lib ones?

dillon-cullinan replied:
Deferring to @harrism

harrism replied:
It needs cudf. It doesn't use nvstrings itself but cudf probably depends on it.

kkraus14 replied:
Shouldn't it only need `libnvstrings` and `libcudf`?

harrism replied:
It needs cudf bindings in the Python API, I assume.

raydouglass commented on conda/recipes/cuspatial/meta.yaml:
```diff
@@ -0,0 +1,49 @@
+    - numba >=0.45.1
+    - dlpack
+    - pyarrow 0.14.1.*
+    - libcudf {{ version }}
```
No dependency on `cudf`?

dillon-cullinan replied:
Since we are building cudf/libcudf from source as part of the conda build process, I'm assuming it should not be part of the dependency list.

kkraus14 replied:
It will eventually, but not quite yet.

raydouglass commented on conda/recipes/libcuspatial/meta.yaml:
```diff
@@ -0,0 +1,54 @@
+    - cmake >=3.12.4
+  host:
+    - librmm {{ minor_version }}.*
+    - libnvstrings {{ minor_version }}.*
```
No dependency on `libcudf`?

dillon-cullinan replied:
See above comment.

rlratzel commented on build.sh:
```diff
@@ -0,0 +1,120 @@
+"
+LIBCUSPATIAL_BUILD_DIR=${REPODIR}/cpp/build
+CUSPATIAL_BUILD_DIR=${REPODIR}/python/cuspatial/build
+BUILD_DIRS="${LIBNVSTRINGS_BUILD_DIR} ${NVSTRINGS_BUILD_DIR} ${LIBCUDF_BUILD_DIR} ${CUDF_BUILD_DIR} ${DASK_CUDF_BUILD_DIR}"
```
BUILD_DIRS looks like it needs to be updated

rlratzel commented:
> I'm also unclear on how needing to build cudf from source might affect this. Will the configure and build of cudf from source happen automatically, or does there need to be something that explicitly calls cudf's build.sh?

I now see that the `gpu`, recipes, etc. scripts call cudf's build.sh, but it seems like the root build.sh should take care of that instead, to be consistent.

dillon-cullinan commented:
> > I'm also unclear on how needing to build cudf from source might affect this. Will the configure and build of cudf from source happen automatically, or does there need to be something that explicitly calls cudf's build.sh?
> 
> I now see that the `gpu`, recipes, etc. scripts call cudf's build.sh, but it seems like the root build.sh should take care of that instead, to be consistent.

cuSpatial requiring a cuDF source build is only temporary. Once they are able to use a conda packages, all cudf build artifacts and source build references will be removed and replaced with the cudf conda package and added to the dependency list for conda.

rlratzel commented:
> > > I'm also unclear on how needing to build cudf from source might affect this. Will the configure and build of cudf from source happen automatically, or does there need to be something that explicitly calls cudf's build.sh?
> > 
> > 
> > I now see that the `gpu`, recipes, etc. scripts call cudf's build.sh, but it seems like the root build.sh should take care of that instead, to be consistent.
> 
> cuSpatial requiring a cuDF source build is only temporary. Once they are able to use a conda packages, all cudf build artifacts and source build references will be removed and replaced with the cudf conda package and added to the dependency list for conda.

Yeah, that's why I was thinking you have only the root build.sh take care of that so you don't have to rip out all the other calls from the other scripts when it changes. Not a big deal at this point if leaving it as-is is easier for you.

harrism commented on build.sh:
```diff
@@ -0,0 +1,120 @@
+HELP="$0 [clean] [libcuspatial] [cuspatial] [-v] [-g] [-n] [-h]
+   clean        - remove all existing build artifacts and configuration (start
+                  over)
+   libcuspatial - build the nvstrings C++ code only
```
```suggestion
   libcuspatial - build the cuSpatial C++ code only
```

harrism commented on build.sh:
```diff
@@ -0,0 +1,120 @@
+   clean        - remove all existing build artifacts and configuration (start
+                  over)
+   libcuspatial - build the nvstrings C++ code only
+   cuspatial    - build the nvstrings Python package
```
```suggestion
   cuspatial    - build the cuSpatial Python package
```

kkraus14 commented on build.sh:
```diff
@@ -0,0 +1,120 @@
+if (( ${NUMARGS} == 0 )) || hasArg libcuspatial; then
+
+    mkdir -p ${LIBCUSPATIAL_BUILD_DIR}
+    cd ${LIBCUDF_BUILD_DIR}
```
Missed copy paste?

```suggestion
    cd ${LIBCUSPATIAL_BUILD_DIR}
```

kkraus14 commented on build.sh:
```diff
@@ -0,0 +1,120 @@
+fi
+
+# Build and install the cuspatial Python package
+if (( ${NUMARGS} == 0 )) || hasArg cudf; then
```
```suggestion
if (( ${NUMARGS} == 0 )) || hasArg cuspatial; then
```

kkraus14 commented on ci/checks/changelog.sh:
```diff
@@ -0,0 +1,40 @@
+#!/bin/bash
+# Copyright (c) 2018, NVIDIA CORPORATION.
+#########################
+# cuDF CHANGELOG Tester #
```
```suggestion
# cuSpatial CHANGELOG Tester #
```

kkraus14 commented on ci/checks/style.sh:
```diff
@@ -0,0 +1,40 @@
+#!/bin/bash
+# Copyright (c) 2018, NVIDIA CORPORATION.
+#####################
+# cuDF Style Tester #
```
```suggestion
# cuSpatial Style Tester #
```

kkraus14 commented on ci/checks/style.sh:
```diff
@@ -0,0 +1,40 @@
+FLAKE_RETVAL=$?
+
+# Run flake8-cython and get results/return code
+FLAKE_CYTHON=`flake8 --config=python/cudf/.flake8.cython`
```
```suggestion
FLAKE_CYTHON=`flake8 --config=python/cuspatial/.flake8.cython`
```

kkraus14 commented on ci/cpu/build.sh:
```diff
@@ -0,0 +1,67 @@
+# Set path and build parallel level
+export PATH=/conda/bin:/usr/local/cuda/bin:$PATH
+export PARALLEL_LEVEL=4
+export CUDF_HOME="${WORKSPACE}/cudf"
```
```suggestion
export CUSPATIAL_HOME="${WORKSPACE}/cuspatial"
```

kkraus14 commented on ci/cpu/libcuspatial/build_libcuspatial.sh:
```diff
@@ -0,0 +1,5 @@
+set -e
+
+echo "Building libcuspatial"
+conda build conda/recipes/libcuspatial --python=$PYTHON
```
Does this depend on a Python version? It shouldn't as far as I know:
```suggestion
conda build conda/recipes/libcuspatial
```

dillon-cullinan replied:
I believe it actually does for now. libcuspatial builds libnvstrings, nvstrings, libcudf and cudf as part of its conda build.

kkraus14 replied:
With #39 it won't 😄

kkraus14 commented on ci/cpu/prebuild.sh:
```diff
@@ -0,0 +1,16 @@
+#!/usr/bin/env bash
+
+#Upload cuspatial once per CUDA
```
Only upload on CUDA 9.2 so it uploads on both Python 3.6 and 3.7
```suggestion
#Upload cuspatial once per PYTHON
```

kkraus14 commented on ci/cpu/prebuild.sh:
```diff
@@ -0,0 +1,16 @@
+    export UPLOAD_CUSPATIAL=0
+fi
+
+#Upload libcuspatial once per PYTHON
```
Only upload on Python 3.6 so it uploads on both CUDA 9.2 and CUDA 10.0
```suggestion
#Upload libcuspatial once per CUDA
```

kkraus14 commented on ci/cpu/upload_anaconda.sh:
```diff
@@ -0,0 +1,43 @@
+    return 0
+fi
+
+if [ "$UPLOAD_LIBCUDF" == "1" ]; then
```
```suggestion
if [ "$UPLOAD_LIBCUSPATIAL" == "1" ]; then
```

dillon-cullinan commented:
I believe this is good to merge along with https://github.com/rapidsai/gpuci-scripts/pull/60.

GitHub webhook was added to the repo. CPU tests passed. Cannot test GPU builds as a GPU build is not available on dev Jenkins, so we can check those on another PR and if there are issues with it I can open another PR to fix those problems.

kkraus14 commented on ci/cpu/upload_anaconda.sh:
```diff
@@ -0,0 +1,43 @@
+
+set -e
+
+export LIBCUSPATIAL_FILE=`conda build conda/recipes/libcuspatial --python=$PYTHON --output`
```
```suggestion
export LIBCUSPATIAL_FILE=`conda build conda/recipes/libcuspatial --output`
```

kkraus14 commented on ci/cpu/upload_anaconda.sh:
```diff
@@ -0,0 +1,43 @@
+#!/bin/bash
+#
+# Adopted from https://github.com/tmcdonell/travis-scripts/blob/dfaac280ac2082cd6bcaba3217428347899f2975/update-accelerate-buildbot.sh
```
I don't really think our script is based off this anymore, remove?

kkraus14 commented on ci/cpu/build.sh:
```diff
@@ -0,0 +1,70 @@
+conda config --set ssl_verify False
+
+##########################################################################################
+# BUILD - Conda package builds (conda deps: libcupatial <- libcuspatial_cffi <- cuspatial)
```
```suggestion
# BUILD - Conda package builds (conda deps: libcupatial <- cuspatial)
```

kkraus14 commented on conda/recipes/libcuspatial/build.sh:
```diff
@@ -0,0 +1,10 @@
+# build cuspatial with verbose output
+cd $WORKSPACE
+echo "Fix permissions on build script"
+chmod ugo+x ./build.sh
```
Why is this needed? Should the permissions be fixed on the file and then committed to the repo instead?

kkraus14 commented on ci/gpu/build.sh:
```diff
@@ -0,0 +1,94 @@
+source activate gdf
+
+# These installs are dependencies of cudf and most should be removed once we use a conda package
+# for cudf
```
Stale comment now?

kkraus14 commented on ci/gpu/build.sh:
```diff
@@ -0,0 +1,94 @@
+
+# These installs are dependencies of cudf and most should be removed once we use a conda package
+# for cudf
+conda install "rmm=$MINOR_VERSION.*" "cudf=$MINOR_VERSION" "cudatoolkit=$CUDA_REL"
```
Be consistent with minor version usage.

kkraus14 replied:
We install rmm explicitly here, but don't in the conda recipes. Let's be consistent about it.

kkraus14 commented on ci/gpu/build.sh:
```diff
@@ -0,0 +1,94 @@
+conda list
+
+################################################################################
+# BUILD - Build libnvstrings, nvstrings, libcudf, and cuDF from source
```
Stale comment

kkraus14 commented on ci/gpu/build.sh:
```diff
@@ -0,0 +1,94 @@
+export PARALLEL_LEVEL=4
+export CUDA_REL=${CUDA_VERSION%.*}
+export CUDF_HOME="${WORKSPACE}/cudf"
+export PYTHONPATH=${WORKSPACE}/python/cuspatial
```
We shouldn't do this anymore.
```suggestion
```
